src/librustc_typeck/check/mod.rs

```diff
@@ -3763,8 +3763,36 @@ pub fn resolve_ty_and_def_ufcs<'a, 'b, 'tcx>(fcx: &FnCtxt<'b, 'tcx>,
                                                         &'a [ast::PathSegment],
                                                         def::Def)>
 {
+
+    // Associated constants can't depend on generic types.
+    fn have_disallowed_generic_consts<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
+                                                def: def::Def,
+                                                ty: Ty<'tcx>,
+                                                span: Span,
+                                                node_id: ast::NodeId) -> bool {
+        match def {
+            def::DefAssociatedConst(..) => {
+                if ty::type_has_params(ty) || ty::type_has_self(ty) {
```

Comment:
This is a temporary restriction to prevent ICEs like #25046, until something like rust-lang/rfcs#1062 is implemented. Right now we expect to be able to evaluate constants in `typeck`, but if the constant's value depends on a type parameter, it can't be evaluated until `trans`. To deal with generic constants, we use have something like a projection, but this is much easier to implement in the meantime.

The only reason that a type parameter can affect the value of a const seems to be because it can affect which const we actually resolve to. This doesn't seem to be true of lifetime parameters, so I don't think that we need to worry about them. (And this makes some sense, because constants are... constant, more like statics or macros than something that's restricted to a particular region.) In any case, I haven't been able to construct an equivalent of #25046 that crashes due to a rogue lifetime parameter.